src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -245,6 +349,95 @@ private Exception GetCurlException(int code)
             return new Exception(GetCurlErrorString(code));
         }
 
+        private void SetCurlOption(SafeCurlHandle handle, int option, object value)
+        {
+            int result;
+
+            // TODO: For some reason using generic is not working here
+            //       SetCurlOption<TOption>(SafeCurlHandle, int, TOption) compilation fails
+            if (value is string)
+            {
+                result = Interop.libcurl.curl_easy_setopt(handle, option, (string)value);
+            }
+            else if (value is long)
+            {
+                result = Interop.libcurl.curl_easy_setopt(handle, option, (long)value);
+            }
+            else if (value is Interop.libcurl.curl_readwrite_callback)
+            {
+                result = Interop.libcurl.curl_easy_setopt(handle, option, (Interop.libcurl.curl_readwrite_callback)value);
+            }
+            else if (value is Interop.libcurl.curl_ioctl_callback)
+            {
+                result = Interop.libcurl.curl_easy_setopt(handle, option, (Interop.libcurl.curl_ioctl_callback)value);
+            }
+            else
+            {
+                result = Interop.libcurl.curl_easy_setopt(handle, option, (IntPtr) value);
+            }
+
+            if (result != CURLcode.CURLE_OK)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+            }
+        }
+
+        private static bool InitializeCurl()
+        {
+            // TODO: A process-wide curl_global_init is required. According to the documentation,
+            //       this should be done only once for libcurl and its dependencies (eg. SSL) before
+            //       creating threads that use it. Similarly curl_global_cleanup needs to be called
+            //       when there are no more threads (present or future) requiring libcurl usage.
+            int result = Interop.libcurl.curl_global_init(Interop.libcurl.CurlGlobalFlags.CURL_GLOBAL_ALL);
+            if (result != CURLcode.CURLE_OK)
+            {
+                throw new InvalidOperationException("Cannot use libcurl in this process");
+            }
+            return true;
+        }
+
+        private IntPtr SetRequestHeaders(SafeCurlHandle handle, RequestCompletionSource state)
+        {
+            HttpRequestMessage request = state.RequestMessage;
```

Comment:
If we need only `RequestMessage` from `state` in this method, consider pass it directly  instead of `RequestCompletionSource`